File: leetcode_problems/greedy_algorithms/lc_45_jump_game_II.cpp
/**
 * Leetcode-45: Jump Game - II
 * Link: https://leetcode.com/problems/jump-game-ii/description/
*/

#include <bits/stdc++.h>
using namespace std;

int fn_r(vector<int> &nums, int ind, int cnt){
    if(ind >= nums.size()-1)
        return cnt;

    if(nums[ind] == 0)
        return INT_MAX;
    
    int mini = nums.size() + 1;
    for(int i=1; i <= nums[ind]; i++){
        mini = min(mini, fn_r(nums, ind+i, cnt+1));
    }
    // cout << mini << endl;
    return mini;
}

int jump_rec(vector<int>& nums) {
    int cnt = 0, ind = 0;
    return fn_r(nums, ind, cnt);
}


int jump(vector<int>& nums){
    int jumps = 0, l=0, r=0;
    while(r < nums.size()-1){
        int farthest = 0;
        for(int i=l; i<=r; i++){
            farthest = max(farthest, i+nums[i]);
        }
        l = r+1;
        r = farthest;
        jumps ++;
    }
    return jumps;
}




int main(){
    vector<int> nums{2,3,1,1,4};
    cout << jump(nums) << endl;
}

File: leetcode_rs/greedy_algorithms/lc_45_jump_game_II.rs
/**
 * Leetcode-45: Jump Game - II
 * Link: https://leetcode.com/problems/jump-game-ii/description/
*/



// This solution is a Brute Force -> Trying out all possible using recursion

fn fn_r(nums: &Vec<i32>, ind: i32, cnt: i32) -> i32 {
    if ind >= (nums.len() - 1) as i32 {
        return cnt;
    }

    let mut mini: i32 = i32::MAX;
    let mut i: i32 = 1;
    while i <= nums[ind as usize] {
        mini = std::cmp::min(mini, fn_r(&nums, i as i32 + ind, cnt+1));
        i += 1;
    }
    return mini;
}

fn min_jumps(nums: &Vec<i32>) -> i32{
    let ind: i32 = 0;
    let cnt: i32 = 0;
    return fn_r(&nums, ind, cnt);
}



// Optimal Solution
fn min_jumps_optimal(nums: &Vec<i32>) -> i32{
    let mut jumps: i32 = 0;
    let mut l : i32 = 0;
    let mut r: i32 = 0;

    while r < (nums.len()-1) as i32 {
        let mut farthest: i32 = 0;
        for i in l..r+1 {
            farthest = std::cmp::max(farthest, i+nums[i as usize]);
        }
        l = r+1;
        r = farthest;
        jumps += 1;
    }
    return jumps;
}


fn main(){
    let nums: Vec<i32> = vec![2,3,1,4,1,1,1,2];
    println!("{}", min_jumps_optimal(&nums));
}